redis/query_utils.cc

```diff
@@ -70,22 +71,25 @@ class strings_result_builder {
 };
 
 future<lw_shared_ptr<strings_result>> read_strings(service::storage_proxy& proxy, const redis_options& options, const bytes& key, service_permit permit) {
-    auto schema = get_schema(proxy, options.get_keyspace_name(), redis::STRINGs);
-    auto ps = partition_slice_builder(*schema).build();
-    query::read_command cmd(schema->id(), schema->version(), ps, 1, gc_clock::now(), std::nullopt, 1); 
-    auto pkey = partition_key::from_single_value(*schema, key);
-    auto partition_range = dht::partition_range::make_singular(dht::global_partitioner().decorate_key(*schema, std::move(pkey)));
-    dht::partition_range_vector partition_ranges;
-    partition_ranges.emplace_back(std::move(partition_range));
-    auto read_consistency_level = options.get_read_consistency_level();
     db::timeout_clock::time_point timeout = db::timeout_clock::now() + options.get_read_timeout();
-    return proxy.query(schema, make_lw_shared(std::move(cmd)), std::move(partition_ranges), read_consistency_level, {timeout, permit, service::client_state::for_internal_calls()}).then([ps, schema] (auto qr) {
-        return query::result_view::do_with(*qr.query_result, [&] (query::result_view v) {
-            auto pd = make_lw_shared<strings_result>();
-            v.consume(ps, strings_result_builder(pd, schema, ps));
-            return pd;
+    return redis::with_strings_locked_key(redis::get_locks_context(options.db_index()), key, timeout, [timeout, &proxy, &options, &key, permit] () mutable {
```

Comment:
I now I understand that the commit message slightly underplays what is being changed here.

As far as I understand now, what you're changing here doesn't **just** effect pipelining in a single request. 
Even if you have multiple requests from multiple connections, this "lock" serializes them so only one happens at a time. Is this intentional?

Although it sounds like this serialization is a good idea in general, but there's an important catch... It only happens on a single shard (the lock manager is thread-local).... So concurrent updates to the same key that happened to have been received on different shards (or even different nodes!) will happen in parallel, without the lock protecting anything. This is good enough for pipelining (since the same request belongs to a single shard) but does it offer anything more?

So given that this global - but per-shard - lock doesn't protect much besides pipelining, why do we make it global at all and not a per-request map, used just for pipelining?
Is it because this way is easier? Or there's something actually better in this approach?

In any case should probably have a comment above the "lock manager" explaining what exactly it tries to serialize, what it doesn't serialize (write on different shards) and why it does what we want related to pipelining.

Reply:
> As far as I understand now, what you're changing here doesn't just effect pipelining in a single request.
> Even if you have multiple requests from multiple connections, this "lock" serializes them so only one happens at a time. Is this intentional?

Sorry, I should write more detail about the serialization of processing requests. It's better to implement the serialization in a separate path. Multiple requests form multiple connections with same key should be processed in order. Obviersly the current implemention is not as expected. It's **not** avoid to update same key concurrently on different shareds.

> So given that this global - but per-shard - lock doesn't protect much besides pipelining, why do we make it global at all and not a per-request map, used just for pipelining?
> Is it because this way is easier? Or there's something actually better in this approach?

As mentioned above, the implemention of serialization of processing requests in this patch is totally wrong. Perhaps there is a better approach. Requests of same key processed by only one shard of the node should be guaranteed. And these Requests of same key should be processed in order. In fact, the requests should be forwarded to the target shard based on the hash value of the key associated. And the requests recieved by the target shard are processed
serially with same key(protected by lock), otherwise are precessed parallel.

So it's better than the one with global locks ?